src/transformers/cache_utils.py

```diff
@@ -1087,8 +1087,8 @@ def update(
         cache_position = cache_kwargs.get("cache_position")
         self.key_cache[layer_idx] = self.key_cache[layer_idx].to(device=key_states.device)
         self.value_cache[layer_idx] = self.value_cache[layer_idx].to(device=value_states.device)
-        k_out = self.key_cache[layer_idx]
-        v_out = self.value_cache[layer_idx]
+        k_out = self.key_cache[layer_idx].clone()
+        v_out = self.value_cache[layer_idx].clone()
```

Comment:
Are the values properly updated (self.key_cache and self.value_cache) since you are doing a copy ? Also do you see a memory increase from this modification ?

Reply:
ah just commented here https://github.com/pytorch/pytorch/issues/131679#issuecomment-2329723501 - Tugsuu is (tentatively) going to look into making that error less intrusive so this case should "just work" without a clone

Reply:
@bdhirsh : thank you so much for looking into this. In a meanwhile we are considering 2 workarounds on HF side:
1. This PR (#33179)
2. PR from @guangy10 (#33287)

@bdhirsh, can you guide us whether 1st or 2nd WA is preferable or you just recommend to wait couple days for the fix on PyTorch side?

Reply:
I'm actually a bit confused by the code. For example - in https://github.com/huggingface/transformers/pull/33287/files, I don't think that code will do anything:
```
# let's say self.key_cache is a tensor with device =='cuda'
# and let's say key_states.device == 'cpu'
self.key_cache[0] = self.key_cache[0].to(device=key_states.device)
```

the above code won't actually change the device of `self.key_cache` (it will remain on cuda, even if `key_states lives on cpu). `self.key_cache[0]` is taking a view/slice off of `self.key_cache`, and you can't change the device of "part" of a tensor.

Reply:
I guess my meta question is: in general, don't you know the device of your kv cache ahead of time? Can you ensure that your model (and its kv cache) are initialized on the right device to begin with, so that when you export, you don't need to do any device conversions? You probably don't want any device conversions in the runtime path (inside of your exported program) anyway, since they will hurt perf

Reply:
> I guess my meta question is: in general, don't you know the device of your kv cache ahead of time?

https://github.com/huggingface/transformers/pull/33303 🤗